feat: add German IPA mapping (deu)

leonhard-behr opened the pull request:
Pull request template for adding a new language
-----------------------------------------------

<!--
  - Use this template if you are adding a new language or modifying a mapping
  - For all other PRs, please add  ?expand=1&template=software_pr.md  to the URL above
    to use the software PR template instead.
 -->

* **Please check if the PR fulfills these requirements**
- [x] Mapping files are added in `g2p/mappings/langs`
- [x] Mapping is either added to an existing folder or a new folder has been added
- [x] Language folder and files use appropriate [ISO 639-3 codes](https://en.wikipedia.org/wiki/List_of_ISO_639-3_codes)
- [x] `config-g2p.yaml` file includes all author names, and settings necessary
- [x] Please add some test data in `g2p/tests/public/data`. The added file should be a csv/tsv/psv file and each row should have the format `[input_mapping_code,output_mapping_code,input_string,output_string]`
- [x] As the last step, G2P has been updated by running `g2p update` locally and committing the change
- [x] You agree to license your contribution under the same license as this project (see [LICENSE](https://github.com/roedoejet/g2p/blob/main/LICENSE) file).

* **Other information**:

The German language mapping was created by me as part of the Project TTS-De at the University of Applied Sciences Ansbach.

leonhard-behr commented on g2p/mappings/langs/deu/deu_to_ipa.csv:
```diff
@@ -0,0 +1,121 @@
+e,ə,,\b
+e,e,KONSONANT,
+ə,e,KONSONANT,
+dɛr,de:ɐ̯,,\b
```
Hey! Thank you for pointing out this bug, it seems to be a typing error. The ascii colon was placed incorrectly instead of the U+02D0 colon.

leonhard-behr commented on g2p/mappings/langs/deu/deu_to_ipa.csv:
```diff
@@ -0,0 +1,121 @@
+ie,iː,,
+iə,iː,,
+ih,iː,,
+ig,iː,\b,
+ig,ɪç,,\b
```
You are right, the ig rules are both not applied in the mapping. The phonetic conversion is correct when placing them before the ``g`` mappings.
During the ``ig`` testing I noticed a small issue with ``œ`` and ``ø``. Those two need to be fixed. I'll see what i can do. When fixed I'll update the mapping table :)

leonhard-behr replied:
The issues should be resolved, i commited the latest changes. I hope that the table is now free of any faulty rules or circular mappings... Thank you for pointing out these bugs! :)

joanise reviewed (approved):
This looks good to me now, and will be ready to merge once I fix the merge conflicts.

joanise commented:
Oh, there is one more question: @dhdaines asks in reviewing #440 the following question:

> why are ö, œ, øː, and œː in BACK? they are front vowels

I think I agree with him. My German isn't fluent anymore, but I think `möchte` should map to `møːçtə` but it currently maps to `møːχtə`.

While testing this, I tried to convert "Ich möchte Sachen." and I think Sachen is mapped incorrectly too:
```
$ g2p convert "Ich möchte Sachen." deu deu-ipa
ɪç møːχtə zaːçən.
```
Should it not be `zaːχən`?

leonhard-behr commented:
> Oh, there is one more question: @dhdaines asks in reviewing #440 the following question:
> 
> > why are ö, œ, øː, and œː in BACK? they are front vowels
> 
> I think I agree with him. My German isn't fluent anymore, but I think `möchte` should map to `møːçtə` but it currently maps to `møːχtə`.
> 
> While testing this, I tried to convert "Ich möchte Sachen." and I think Sachen is mapped incorrectly too:
> 
> ```
> $ g2p convert "Ich möchte Sachen." deu deu-ipa
> ɪç møːχtə zaːçən.
> ```
> 
> Should it not be `zaːχən`?

Hey! Thanks for poining out this error. You are right, ö, œ, ø and their long versions were wrongly listed as BACK. 

In German ``möchte`` is pronounced ``ˈmœçtə``. In the sentence ``ɪç møːχtə zaːçən.`` the phonetics ``ø`` and ``χ`` in `*möchte* are wrong as well as the ``ç`` in ``zaːçən``. 
With updated rules and abbs i correctly get ``ɪç ˈmœçtə zaːχən.``. 


One thing I noticed: i previously made a mistake in the test file: The phonemes of *Sprache* should be ``ʃpʁaːχə`` with a voiceless uvular fricative ``χ`` instead of the current ``ç``. I apologize for the oversight in the testing cases. 

I will review the rules and the testing cases once again to make sure that there are hopefully no more bugs and will commit the changes after my review.

leonhard-behr commented:
I fixed the rule ordering as best as I could, the `ch` and the `ö` section should be working better now. I added `möchte` and `Sachen` to the test cases. 

Thank you for pointing out the bugs and therefore improving the mapping :)

I hope that the main issues are now dealt with. However this mapping is surely not perfect, it is merely an approximation and could possibly be extended even more.
